imbi/endpoints/gitlab.py

```diff
@@ -107,7 +122,50 @@ async def create_project(self, parent, project_name, **attributes) -> dict:
             return response.body
         raise errors.InternalServerError(
             'failed to create project %s: %s', project_name, response.code,
-            title='GitLab API Failure')
+            title='GitLab API Failure', gitlab_response=response.body)
+
+    async def commit_tree(self, project_info: dict, project_dir: pathlib.Path,
+                          commit_message: str):
+        self.logger.info('creating commit for %s', project_info['id'])
+
+        files = [path for path in project_dir.glob('**/*') if path.is_file()]
+        self.logger.debug('found %d files in %s', len(files), project_dir)
+
+        actions = [
+            {
+                'action': 'create',
+                'content': base64.b64encode(file.read_bytes()).decode('ascii'),
+                'encoding': 'base64',
+                'file_path': str(file.relative_to(project_dir)),
+            }
+            for file in files
+        ]
+        actions.extend([
+            {
+                'action': 'chmod',
+                'execute_filemode': True,
+                'file_path': str(file.relative_to(project_dir)),
+            }
+            for file in files
+            if file.stat().st_mode & stat.S_IXUSR
+        ])
+        self.logger.debug('creating commit with %d actions', len(actions))
+
+        project_url = yarl.URL(project_info['_links']['self'])
+        response = await self.api(
+            project_url / 'repository' / 'commits',
+            method='POST',
+            body={
+                'branch': project_info['default_branch'],
+                'commit_message': commit_message,
+                'actions': actions,
+            },
+        )
+        if response.ok:
+            return response.body
+        raise errors.InternalServerError(
+            'failed to commit to %s: %s', project_url, response.code,
```

Comment:
Is there a reason we don't use f strings here? I'm under the impression that f strings are always preferred unless there's a specific reason not to.